pkg/core/chain/chain.go

```diff
@@ -200,11 +206,17 @@ func (c *Chain) Listen() {
 			c.provideLastCertificate(r)
 		case r := <-c.getRoundResultsChan:
 			c.provideRoundResults(r)
+		case msg := <-c.acceptBlockChan:
+			if err := c.onAcceptBlock(ctx, msg); err != nil {
```

Comment:
I'd advise to keep the callback over the channel. The reason for this being that the `Synchronizer` relies on `AcceptBlock` being blocking, otherwise it causes a race condition between when a block is being accepted, and when a new one is downloaded and attempted to be passed down. Since we determine whether or not to pass a block down based on the `prevBlock.Height`, it can result in blocks being dropped inadvertently.

Reply:
problem with that is that we now need a `context.Context` which should be injected when the call is done. It is bad practice to save it in the chain and I am not sure how we can add it to the method signature directly within the callback

Reply:
mmm apparently, the advice to not store context can be relaxed a bit (see: https://github.com/golang/go/issues/22602). I will revert the change and keep the context as a field within the `Chain`